src/hooks/useUserOnboardingContext.ts

```diff
@@ -15,54 +15,86 @@ limitations under the License.
+        (...values: T) => ref.current(...values),
+        [],
+    );
 }
 
 export function useUserOnboardingContext(): UserOnboardingContext | null {
     const [context, setContext] = useState<UserOnboardingContext | null>(null);
 
     const cli = MatrixClientPeg.get();
-    const handler = useCallback(async () => {
-        try {
-            const profile = await cli.getProfileInfo(cli.getUserId());
-
-            const myDevice = cli.getDeviceId();
-            const devices = await cli.getDevices();
-
-            const dmRooms = DMRoomMap.shared().getUniqueRoomsWithIndividuals() ?? {};
-            setContext({
-                avatar: profile?.avatar_url ?? null,
-                myDevice,
-                devices: devices.devices,
-                dmRooms: dmRooms,
-            });
-        } catch (e) {
-            logger.warn("Could not load context for user onboarding task list: ", e);
-            setContext(null);
-        }
-    }, [cli]);
-
-    useEventEmitter(cli, ClientEvent.AccountData, handler);
+    const handler = useRefOf(
+        useCallback(async () => {
+            try {
+                let hasAvatar = context?.hasAvatar;
+                if (!hasAvatar) {
+                    const profile = await cli.getProfileInfo(cli.getUserId());
+                    hasAvatar = Boolean(profile?.avatar_url);
+                }
+
+                let hasDevices = context?.hasDevices;
+                if (!hasDevices) {
+                    const myDevice = cli.getDeviceId();
+                    const devices = await cli.getDevices();
+                    hasDevices = Boolean(devices.devices.find(device => device.device_id !== myDevice));
+                }
+
+                let hasDmRooms = context?.hasDmRooms;
+                if (!hasDmRooms) {
+                    const dmRooms = DMRoomMap.shared().getUniqueRoomsWithIndividuals() ?? {};
+                    hasDmRooms = Boolean(Object.keys(dmRooms).length);
+                }
+
+                let hasNotificationsEnabled = context?.hasNotificationsEnabled;
+                if (!hasNotificationsEnabled) {
+                    hasNotificationsEnabled = Notifier.isPossible();
+                }
+
+                if (hasAvatar !== context?.hasAvatar
+                    || hasDevices !== context?.hasDevices
+                    || hasDmRooms !== context?.hasDmRooms
+                    || hasNotificationsEnabled !== context?.hasNotificationsEnabled) {
+                    setContext({ hasAvatar, hasDevices, hasDmRooms, hasNotificationsEnabled });
+                }
+            } catch (e) {
+                logger.warn("Could not load context for user onboarding task list: ", e);
+                setContext(null);
+            }
+        }, [context, cli]),
+    );
+
     useEffect(() => {
-        const handle = setInterval(handler, 2000);
+        cli.on(ClientEvent.AccountData, handler);
+        const handle = setInterval(handler, USER_ONBOARDING_CONTEXT_INTERVAL);
```

Comment:
Using `setInterval` here could lead to stacking of requests, given that the handler runs many async operations in sequence, if getting profile info takes 3s, and getting devices takes 3s then you'll be launching a new call before this one is even done